Jobservice events

kannon92 opened the pull request:
Solve https://github.com/G-Research/armada/issues/1138

Add a redis cache for JobService.

severinson commented on cmd/jobservice/main.go:
```diff
@@ -0,0 +1,54 @@
+
+const CustomConfigLocation string = "config"
+
+func init() {
```
I suggest looking at how the testsuite is implemented. I think that's a cleaner approach. It doesn't require having an init function and so on.

severinson commented on internal/jobservice/configuration/types.go:
```diff
@@ -0,0 +1,20 @@
+	"github.com/G-Research/armada/pkg/client"
+)
+
+type JobServiceConfiguration struct {
```
I think we can re-use the config we already use for the Armada client? The test suite also takes this approach.

kannon92 replied:
Nice find!  I attempted to do this.  Please take a look.

severinson commented on cmd/jobservice/main.go:
```diff
@@ -0,0 +1,54 @@
+func main() {
+	common.ConfigureLogging()
+	common.BindCommandlineArguments()
+	g, ctx := errgroup.WithContext(context.Background())
```
Why both an errgroup and a waitgroup? I suggest using only errgroups.

severinson commented on cmd/jobservice/main.go:
```diff
@@ -0,0 +1,54 @@
+		}
+	})
+
+	wg.Wait()
```
This doesn't look right. Starting the service in an errgroup and waiting on the waitgroup.

severinson commented on internal/jobservice/application.go:
```diff
@@ -0,0 +1,51 @@
+	grpcServer := grpcCommon.CreateGrpcServer(config.Grpc.KeepaliveParams, config.Grpc.KeepaliveEnforcementPolicy, []authorization.AuthService{&authorization.AnonymousAuthService{}})
+	log.Info("JobService service listening on ", config.GrpcPort)
+
+	redisJobRepository := repository.NewRedisJobServiceRepository(db)
```
Is Redis a good fit for this? Why do we need a RedisJobServiceRepository?

severinson replied:
I'd probably use postgres instead of Redis to store the cache state on disk. And then also store the state of all jobs watched in-memory in the application itself as a map.

severinson commented on internal/jobservice/eventstojobs/event_job_response.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/G-Research/armada/pkg/api/jobservice"
+)
+
+func EventsToJobResponse(message api.EventMessage) (*jobservice.JobServiceResponse, error) {
```
Why do we need this? Are not just interested in tracking the current state of each job?

kannon92 replied:
In short, the job response only cares about terminal events and running.  So we only return those but there are 14-15 other events that can happen so we use this to translate events into our jobresponse.

severinson commented on internal/jobservice/eventstojobs/eventstojobs.go:
```diff
@@ -0,0 +1,72 @@
+		jobServiceRepository: jobServiceRepository}
+}
+
+func (eventToJobService *EventsToJobService) SubscribeToJobSetId(context context.Context) error {
```
I'd use the approach used in the test suite to watch events.

richscott commented on internal/jobservice/eventstojobs/eventstojobs.go:
```diff
@@ -0,0 +1,129 @@
+			return err
+		}
+		for {
+			msg, err := stream.Recv()
```
I sort of wish that protoc generated code such that GetJobSetEvents() returned a Go channel that we could select upon, instead of doing repeated Recv() calls. (This is gogo's protoc? I wonder if other, newer protoc's would generate more Go-chan-oriented code?)

kannon92 replied:
Yea I think we need to do some research into go and protoc anyway.  We can add this to that one issue.

kannon92 commented on config/jobservice/config.yaml:
```diff
@@ -0,0 +1,21 @@
+grpcPort: 60003
+httpPort: 8090
+SubscribeJobSetTime: 10
```
Fixed.

kannon92 commented on internal/jobservice/eventstojobs/eventstojobs.go:
```diff
@@ -0,0 +1,129 @@
+			log.Infof("key %s element: %s", key, element.State)
+			e := eventToJobService.jobServiceRepository.UpdateJobServiceDb(key, element)
+			if e != nil {
+				panic(e)
```
@severinson and @d80tb7 What are your thoughts on error handling in this code base? 

I changed this to return an error but I don't really know what pattern to follow.

jayofdoom commented on internal/jobservice/eventstojobs/event_job_response.go:
```diff
@@ -0,0 +1,56 @@
+	"fmt"
+
+	"github.com/G-Research/armada/pkg/api"
+	"github.com/G-Research/armada/pkg/api/jobservice"
```
I have the opposite preference here -- I don't like aliased package names in golang unless they are required to ensure separated namespaces. I think that makes @kannon92 the tie-breaking vote :laughing:

kannon92 replied:
Since I changed it already, I vote for js.

d80tb7 commented on internal/jobservice/eventstojobs/event_job_response.go:
```diff
@@ -0,0 +1,56 @@
+}
+func IsEventTerminal(message api.EventMessage) bool {
+	switch message.Events.(type) {
+	case *api.EventMessage_Cancelled, *api.EventMessage_Succeeded, *api.EventMessage_Failed:
```
I think DuplicateFound is technically a terminal state

d80tb7 commented on internal/jobservice/server/server.go:
```diff
@@ -0,0 +1,52 @@
+		return nil, err
+	}
+	var subscribeToJobSet = eventstojobs.IsStateTerminal(response.State)
+	if !subscribeToJobSet {
```
oh- I see- you return earlier in the function if that's the case- I stil think you have a race condition though where two people subscribe at the same time.  I think you need a lock.

d80tb7 commented on internal/jobservice/eventstojobs/eventstojobs.go:
```diff
@@ -0,0 +1,129 @@
+				return nil
+			default:
+			}
+			if !IsEventAJobResponse(*msg.Message) {
```
wouldn't this be simpler as just 
```
if IsEventAJobResponse{
  ...
}
```
rather than declaring an empty if and then using the else for the logic

d80tb7 commented on internal/jobservice/eventstojobs/eventstojobs.go:
```diff
@@ -0,0 +1,129 @@
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			// This allows us to subscribe for x amount of time
```
why do we want to subscribe for some time?

d80tb7 commented on internal/jobservice/eventstojobs/eventstojobs.go:
```diff
@@ -0,0 +1,129 @@
+					terminalEventClientId = IsEventTerminal(*msg.Message)
+				}
+				val, ok := jobIdMap[currentJobId]
+				if ok && val.State != jobStatus.State {
```
why do we have an if here when we do the same on both conditions?

kannon92 commented on internal/jobservice/eventstojobs/eventstojobs.go:
```diff
@@ -0,0 +1,129 @@
+	return &EventsToJobService{
+		queue:                queue,
+		jobsetid:             jobsetid,
+		jobid:                jobid,
```
We can discuss.

In general, I found that I can't really subscribe from a client grpc call in the way that you would prefer.  If you return a rpc call, it will cancel the go routine once the call returns.  

So I built a subscribe timer that subscribes on a job-set for x amount of time and it will return if you have a terminal state for your job id.

kannon92 commented on internal/jobservice/eventstojobs/eventstojobs.go:
```diff
@@ -0,0 +1,129 @@
+	// But we will return after SubscribeJobSetTime s
+	ctxTimeout, cancel := context.WithTimeout(ctx, time.Duration(eventToJobService.jobServiceConfig.SubscribeJobSetTime)*time.Second)
+	defer cancel()
+	err := client.WithEventClient(&eventToJobService.jobServiceConfig.ApiConnection, func(c api.EventClient) error {
```
Good call.

kannon92 commented:
Decided to do this as a new PR.  I addressed the comments of this PR in https://github.com/G-Research/armada/pull/1259 and followed up with Chris offline to discuss the implementation.
